### DISCUSSION TITLE
Airflow 2.2.1; Scheduler heartbeat fails

### DISCUSSION THREAD

**Discussion by sivaramanl** [apache/airflow#22669] [Q&A]
Hello all, I am facing an issue with the Airflow scheduler heartbeat not getting updated regularly. We are using Airflow 2.2.1 in HA mode with MySQL as the backend. Recently, we duplicated the same DAG multiple times to test the schedules and ended up with the issue where the scheduler heartbeat keeps failing. On checking the database, I found that the job table entry for the Scheduler job was getting updated once every 2 - 5 minutes only causing the scheduler check to fail. Could someone provide any pointers to help debug this issue?

**Comment by potiuk** [comment]
Please follow best practices here - https://airflow.apache.org/docs/apache-airflow/stable/best-practices.html#top-level-python-code to avoid running heavy processing in Top-level code of DAGs.

**Reply by potiuk** [reply]
This is know to solve all similar problems.

**Reply by sivaramanl** [reply]
Thank you for your response. In my case, I only have the task definitions in the DAG at the top level. And the DAG was running fine until multiple copies of the same DAG were added to Airflow.

**Reply by sivaramanl** [reply]
I want to add more information. Currently, there are 3850 DAGs in the Airflow environment, and the tasks are getting scheduled, albeit pretty slowly since I have set the parallelism to 16. On increasing the parallelism to 256, I confirmed that more tasks are scheduled in parallel and consumed by the workers. However, the scheduler heartbeat entry in the database is not updated, leading to the liveness probe configured for the scheduler to fail.

This is the scheduler probe currently used:
```
import os
import sys
from airflow.jobs.scheduler_job import SchedulerJob
from airflow.utils.db import create_session
from airflow.utils.net import get_hostname
os.environ['AIRFLOW__CORE__LOGGING_LEVEL'] = 'ERROR'
os.environ['AIRFLOW__LOGGING__LOGGING_LEVEL'] = 'ERROR'

with create_session() as session:
    job = session.query(SchedulerJob).filter_by(hostname=get_hostname()).order_by(SchedulerJob.latest_heartbeat.desc()).limit(1).first()
sys.exit(0 if job.is_alive() else 1)
```

**Reply by potiuk** [reply]
No idea. Have not heard about similar issue. I'd still recommend you to check if you are not doing "too much" at the top-level. Even big and long-time-taking imports might be your problem. 

I suggest you to use things like py-spy and see whether you have no long imports in your tasks.

**Reply by potiuk** [reply]
Also make sure you do not have duplicated dag_ids - they might actually cause some problems like diferent schedulers trying to update same dag_id from different files ? Wild guessing.

**Comment by fackyhigh** [comment]
Hi @sivaramanl. Have you figured out what was the problem? We're experiencing the same issue.
